std/experimental/checkedint.d

```diff
@@ -0,0 +1,1584 @@
+/**
+
+This module defines facilities for efficient checking of integral operations
+against overflow, casting with loss of precision, unexpected change of sign,
+etc. The checking (and possibly correction) can be done at operation level, for
+example $(D opChecked!"+"(x, y, overflow)) adds two integrals `x` and `y` and
+sets `overflow` to `true` if an overflow occurred. The flag (passed by
+reference) is not touched if the operation succeeded, so the same flag can be
+reused for a sequence of operations and tested at the end.
+
+Issuing individual checked operations is flexible and efficient but often
+tedious. The `Checked` facility offers encapsulated integral wrappers that do
+all checking internally and have configurable behavior upon erroneous results.
+For example, `Checked!int` is a type that behaves like `int` but issues an
+`assert(0)` whenever involved in an operation that produces the arithmetically
```

Comment:
> Technically not the same

It's a very important difference:
* `assert(0)` is `@nogc`, `throw new Error("message");` is not.
* `assert(0)` produces terrible cryptic crash messages in release mode (on Linux, at least). `throw new Error("message");` does not.